src/RHI/RHITimestampQueryPool.cpp

```diff
@@ -0,0 +1,161 @@
+﻿#include "RHITimestampQueryPool.h"
+
+#include <Vex/Graphics.h>
+#include <Vex/RHIImpl/RHICommandList.h>
+#include <Vex/Validation.h>
+
+namespace vex
+{
+
+QueryHandle RHITimestampQueryPoolBase::AllocateQuery(QueueType queueType)
+{
+    u32 nextHead = (head + 1) % MaxInFlightQueriesCount;
+    if (nextHead == tail)
+    {
+        u32 processedQueries = ResolveQueries();
+        VEX_CHECK(processedQueries != 0,
+                  "Unable to make room for new timestamp query. Max in flight unresolved queries reached");
+    }
+
+    QueryHandle handle = QueryHandle::CreateHandle(head, generation);
+    inFlightQueries[head].handle = handle;
+    inFlightQueries[head].queueType = queueType;
+    inFlightQueries[head].token = GInfiniteSyncTokens[queueType];
+
+    // advance head
+    u32 prevHead = head;
+    head = nextHead;
+
+    // advance generation
+    if (head < prevHead)
+    {
+        generation++;
+        CleanupOldQueries();
+    }
+
+    return handle;
+}
+
+std::expected<Query, QueryStatus> RHITimestampQueryPoolBase::GetQueryData(QueryHandle handle)
+{
+    if (const auto it = resolvedQueries.find(handle); it != resolvedQueries.end())
+    {
+        return it->second;
+    }
+
+    if (handle.GetGeneration() < generation - 1)
```

Comment:
Does this work correctly for the the first iteration (eg: when generation == 0). I'm guessing generation is a u32.